### PR TITLE
boa: support running Python code in non-blocking way

### PR DESCRIPTION
**Pull request by yorkie**
This PR uses the Node.js `worker_threads` to start a new thread to run Python functions, and add `SharedPythonObject` to share(not copy) the Python object in sub-threads.

The following is an experimental API usage(example):

```js
const { Worker, isMainThread, workerData, parentPort } = require('worker_threads');
const boa = require('@pipcook/boa');
const pybasic = boa.import('tests.base.basic');
const { SharedPythonObject, symbols } = boa;

class Foobar extends pybasic.Foobar {
  hellomsg(x) {
    return `hello <${x}> on ${this.test}(${this.count})`;
  }
}

if (isMainThread) {
  const foo = new Foobar();
  const worker = new Worker(__filename, {
    workerData: {
      foo: new SharedPythonObject(foo),
      // the `foo` is unavailable util the sub-thread exits.
      // it throws an error if trying to use `foo`.
    },
  });
  let alive = setInterval(() => {
    // this is always alive though it's blocking by `foo.sleep()` in sub-thread.
    console.log(`main: still training, ownership(${foo[symbols.GetOwnershipSymbol]()})`);
  }, 1000);

  worker.on('message', (state) => {
    if (state === 'done') {
      clearInterval(alive);
      // the process exits
    }
  });
} else {
  const { foo } = workerData; // the foo is a borrowed object from main thread by `SharedPythonObject`.
  foo.sleep(); // call Python's method `sleep()`, it would block the sub-thread 5s.
  setTimeout(() => {
    // this will 1) do ref the sub-thread 2000ms, 2) send a "done" message and 3) exit the sub-thread.
    parentPort.postMessage('done');
  }, 2000);
}
```

> Note: an object could only be shared with 1 sub-thread at the same time.

### TODO

- [ ] improvements
  - [x] avoid to create `SharedPythonObject` twice and multiple times on a same object.
  - [ ] <s>avoid crash(segfault) if the `ownershipId` is invalid.</s>
- [x] unit tests
- [x] documentation

### CODE REVIEW

**Comment by yorkie** [comment]
@FeelyChau Just mark this PR be ready for review :)

**Comment by yorkie** [comment]
Occurred the following error if running `npm test`:

```
* thread #1, stop reason = signal SIGSTOP
  * frame #0: 0x000000010554b27e libpython3.7m.dylib`PyErr_FormatV + 30
    frame #1: 0x000000010554b239 libpython3.7m.dylib`PyErr_Format + 137
    frame #2: 0x0000000105478590 libpython3.7m.dylib`_PyObject_GenericGetAttrWithDict + 848
    frame #3: 0x0000000105477b26 libpython3.7m.dylib`PyObject_GetAttrString + 134
    frame #4: 0x0000000105477c4c libpython3.7m.dylib`PyObject_HasAttrString + 12
    frame #5: 0x000000010526c589 boa.node`boa::PythonObject::HasAttr(Napi::CallbackInfo const&) [inlined] pybind11::hasattr(obj=<unavailable>, name=<unavailable>) at pytypes.h:397:12 [opt]
    frame #6: 0x000000010526c584 boa.node`boa::PythonObject::HasAttr(this=0x00000001053091e0, info=0x00007ffeefbfc4a0) at object.cc:315 [opt]
    frame #7: 0x0000000105271a9b boa.node`Napi::ObjectWrap<boa::PythonObject>::InstanceMethodCallbackWrapper(this=<unavailable>)::'lambda'()::operator()() const at napi-inl.h:3838:12 [opt]
    frame #8: 0x00000001052719ca boa.node`Napi::ObjectWrap<boa::PythonObject>::InstanceMethodCallbackWrapper(napi_env__*, napi_callback_info__*) [inlined] napi_value__* Napi::details::WrapCallback<Napi::ObjectWrap<boa::PythonObject>::InstanceMethodCallbackWrapper(napi_env__*, napi_callback_info__*)::'lambda'()>(Napi::ObjectWrap<boa::PythonObject>::InstanceMethodCallbackWrapper(napi_env__*, napi_callback_info__*)::'lambda'()) at napi-inl.h:73:12 [opt]
    frame #9: 0x00000001052719c5 boa.node`Napi::ObjectWrap<boa::PythonObject>::InstanceMethodCallbackWrapper(env=0x000000010500b280, info=0x00007ffeefbfc580) at napi-inl.h:3831 [opt]
    frame #10: 0x000000010006346a node`v8impl::(anonymous namespace)::FunctionCallbackWrapper::Invoke(v8::FunctionCallbackInfo<v8::Value> const&) + 122
    frame #11: 0x00000001002580c8 node`v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo) + 616
    frame #12: 0x000000010025765c node`v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) + 524
    frame #13: 0x0000000100256d82 node`v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) + 258
    frame #14: 0x0000000100a50f59 node`Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit + 57
```

**Comment by yorkie** [comment]
> Occurred the following error if running `npm test`:
> 
> ```
> * thread #1, stop reason = signal SIGSTOP
>   * frame #0: 0x000000010554b27e libpython3.7m.dylib`PyErr_FormatV + 30
>     frame #1: 0x000000010554b239 libpython3.7m.dylib`PyErr_Format + 137
>     frame #2: 0x0000000105478590 libpython3.7m.dylib`_PyObject_GenericGetAttrWithDict + 848
>     frame #3: 0x0000000105477b26 libpython3.7m.dylib`PyObject_GetAttrString + 134
>     frame #4: 0x0000000105477c4c libpython3.7m.dylib`PyObject_HasAttrString + 12
>     frame #5: 0x000000010526c589 boa.node`boa::PythonObject::HasAttr(Napi::CallbackInfo const&) [inlined] pybind11::hasattr(obj=<unavailable>, name=<unavailable>) at pytypes.h:397:12 [opt]
>     frame #6: 0x000000010526c584 boa.node`boa::PythonObject::HasAttr(this=0x00000001053091e0, info=0x00007ffeefbfc4a0) at object.cc:315 [opt]
>     frame #7: 0x0000000105271a9b boa.node`Napi::ObjectWrap<boa::PythonObject>::InstanceMethodCallbackWrapper(this=<unavailable>)::'lambda'()::operator()() const at napi-inl.h:3838:12 [opt]
>     frame #8: 0x00000001052719ca boa.node`Napi::ObjectWrap<boa::PythonObject>::InstanceMethodCallbackWrapper(napi_env__*, napi_callback_info__*) [inlined] napi_value__* Napi::details::WrapCallback<Napi::ObjectWrap<boa::PythonObject>::InstanceMethodCallbackWrapper(napi_env__*, napi_callback_info__*)::'lambda'()>(Napi::ObjectWrap<boa::PythonObject>::InstanceMethodCallbackWrapper(napi_env__*, napi_callback_info__*)::'lambda'()) at napi-inl.h:73:12 [opt]
>     frame #9: 0x00000001052719c5 boa.node`Napi::ObjectWrap<boa::PythonObject>::InstanceMethodCallbackWrapper(env=0x000000010500b280, info=0x00007ffeefbfc580) at napi-inl.h:3831 [opt]
>     frame #10: 0x000000010006346a node`v8impl::(anonymous namespace)::FunctionCallbackWrapper::Invoke(v8::FunctionCallbackInfo<v8::Value> const&) + 122
>     frame #11: 0x00000001002580c8 node`v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo) + 616
>     frame #12: 0x000000010025765c node`v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments) + 524
>     frame #13: 0x0000000100256d82 node`v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments, v8::internal::Isolate*) + 258
>     frame #14: 0x0000000100a50f59 node`Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit + 57
> ```

This is addressed, the sub-thread would call to the `finalize_interpreter()` which causes the following calls to Python would crash, just added an extra state `initialized` to flag the finalization op must be in the main thread and `Py_IsInitialized()` is true.

**Comment by yorkie** [comment]
@FeelyChau See https://github.com/alibaba/pipcook/pull/602/checks?check_run_id=1192579536, it seems failed by daemon tests.

**Comment by FeelyChau** [comment]
> @FeelyChau See https://github.com/alibaba/pipcook/pull/602/checks?check_run_id=1192579536, it seems failed by daemon tests which are not related to this PR.

OK, I'll check it out.

**Review by WenheLI — APPROVED** [review]
LGTM

File: packages/boa/src/core/object.h [hunk 1]
```diff
@@ -8,6 +8,27 @@ using namespace Napi;
...
+  }
+  void setOwned(bool owned) {
+    std::lock_guard<std::mutex> l(_mtx);
+    _owned = owned;
+  }
+
+private:
+  mutable std::mutex _mtx;
+  PyObject *_pyobject;
+  bool _owned;
```
Comment (FeelyChau) [inline packages/boa/src/core/object.h:29]: We can use `std::atomic_flag` instead to improve performance.

**Review by FeelyChau — CHANGES_REQUESTED** [review]

**Comment by FeelyChau** [comment]
LGTM.

**Review by FeelyChau — APPROVED** [review]

**Pull request merged**